data-plane/receiver-loom/src/test/java/dev/knative/eventing/kafka/broker/receiverloom/LoomKafkaProducerTest.java

```diff
@@ -0,0 +1,212 @@
+    public void tearDown() {
+        producer.close();
+    }
+
+    @Test
+    public void testConcurrentRecordSendCountAndSequence() throws InterruptedException {
+
+        // Set up the test parameters
+        int numRecords = 100000;
+        AtomicInteger counter = new AtomicInteger();
+        CountDownLatch latch = new CountDownLatch(numRecords);
+        List<Integer> receivedOrder = new ArrayList<>();
+
+        // Send the records
+        for (int i = 0; i < numRecords; i++) {
+            int finalI = i;
+            sendRecord(counter, latch, receivedOrder, finalI);
+        }
+
+        latch.await();
+
+        // Verify the number of records sent
+        assertEquals(numRecords, counter.get());
+
+        // Verify the order of received records
+        for (int i = 0; i < numRecords; i++) {
+            assertEquals(i, receivedOrder.get(i));
+        }
+    }
+
+    @Test
+    public void testSendAfterClose(VertxTestContext testContext) {
+        // Close the producer before sending a record
+        producer.close();
+
+        // Attempt to send a record after the producer is closed
+        ProducerRecord<String, Integer> record = new ProducerRecord<>("test", "sequence number", 123);
+        Future<RecordMetadata> sendFuture = producer.send(record);
+
+        // Verify that the sendFuture fails with the expected error message
+        sendFuture.onComplete(ar -> {
+            testContext.verify(() -> {
+                assertTrue(ar.failed());
+                assertEquals("Producer is closed", ar.cause().getMessage());
+                testContext.completeNow();
+            });
+        });
+    }
+
+    @Test
+    public void testCloseIsWaitingForEmptyQueue(VertxTestContext testContext) {
+        // Send multiple records
+        int numRecords = 100000;
+        List<Future<RecordMetadata>> sendFutures = new ArrayList<>();
+        for (int i = 0; i < numRecords; i++) {
+            ProducerRecord<String, Integer> record = new ProducerRecord<>("test", "sequence number", i);
+            sendFutures.add(producer.send(record));
+        }
+
+        // Close the producer and wait for the event queue to be empty
+        producer.close().onComplete(ar -> {
+            testContext.verify(() -> {
+                assertTrue(ar.succeeded());
+
+                // Ensure that the eventQueue is empty after closing
+                assertTrue(producer.getEventQueueSize() == 0);
+                // Ensure that the sendFromQueueThread is not alive after closing
+                assertTrue(!producer.isSendFromQueueThreadAlive());
+                testContext.completeNow();
+            });
+        });
+
+        // Verify that the sent record is processed before the producer is closed
+        for (int i = 0; i < numRecords; i++) {
+            sendFutures.get(i).onComplete(ar -> {
+                testContext.verify(() -> {
+                    assertTrue(ar.succeeded());
+                    testContext.completeNow();
```

Comment:
Calling `completeNow` multiple times doesn't do what we want here, we expect to have `numRecords + 1 (for close)` "future completions", for that we can use checkpoints:

```java

        final var checkpoints = testContext.checkpoint(numRecords + 1);
        
        testContext.verify(() -> {
                assertTrue(ar.succeeded());
                // other assertions

                checkpoints.flag(); // complete checkpoint
            });

Reply:
So for all of the suggestion we will have the test something like
```java
    @Test
    public void testCloseIsWaitingForEmptyQueue(VertxTestContext testContext) {
        // Send multiple records
        int numRecords = 100000;
        final var checkpoints = testContext.checkpoint(numRecords + 1);

        for (int i = 0; i < numRecords; i++) {
            ProducerRecord<String, Integer> record = new ProducerRecord<>("test", "sequence number", i);
            producer.send(record)
                    .onSuccess(ar -> {
                        checkpoints.flag();
                    })
                    .onFailure(testContext::failNow);
        }

        // Close the producer and wait for the event queue to be empty
        producer.close()
                .onComplete(ar -> {
                    testContext.verify(() -> {
                        assertTrue(ar.succeeded());
                        assertEquals(0, producer.getEventQueueSize());
                        assertFalse(producer.isSendFromQueueThreadAlive());

                        checkpoints.flag();
                    });
                })
                .onFailure(testContext::failNow);
    }
```

Reply:
Yes!